### DISCUSSION TITLE
Can we iteratively stream results in the CLI to stdout ?

### DISCUSSION THREAD

**Discussion by lhoestq** [duckdb/duckdb#12307] [Q&A]
Is it possible to steam the result of a query to stdout (and flushing relatively often) ?

It seems like the CLI waits for the full result to be ready before flushing to stdout (or the flushing chunk size is too big for me to notice).

I'm asking because it could be useful for multiple cases:
- stream data from parquet files to train language models
- stream outputs from a data inspection query to get results in real time (e.g. looking for anomalies)

I tried this command but it flushes stdout once while I'd like the CLI to print iteratively:

```
duckdb -c "COPY (SELECT text FROM 'hf://datasets/HuggingFaceFW/fineweb/sample/10BT/*.parquet' WHERE text SIMILAR TO 'J.*' LIMIT 10) TO '/dev/stdout'WITH (FORMAT 'csv', HEADER);"
```

**Comment by mskyttner** [comment]
When it comes to parquet (if I understand correctly based on https://github.com/duckdb/duckdb/issues/2127), this particular format appears to be difficult to read chunkwise (by design) since the metadata resides at the end of the file.

**Reply by Tishj** [reply]
@mskyttner I don't think that's related

It would be hard to *read* a parquet stream in chunks, but this is outputting the result of the scan to as a CSV, which I think is possible to chunk.

**Reply by lhoestq** [reply]
Parquet can be read row group by row group indeed, but maybe it's possible to get a streamed output anyway ?

For example if there is one row that matches the WHERE clause in the first row group, it can be streamed directly to stfoud without having to wait for results of the remaining row groups ?

**Reply by Tishj** [reply]
Why are we discussing streaming parquet when your example is using `'/dev/stdout' WITH (FORMAT 'csv', HEADER);` ?
Hmm I guess it is reading from parquet, which could have a result on whether it is possible to stream the output

**Reply by lhoestq** [reply]
Yes the question is more on the output: can we get the CSV output streamed to stdout quasi line-by-line rather than having to wait for all the CSV lines to be collected by duckdb ?
